src/main/java/eu/dissco/core/digitalspecimenprocessor/service/AnnotationPublisherService.java

```diff
@@ -0,0 +1,200 @@
+                  .withType(Type.AS_APPLICATION),
+              annotationProcessingRequest));
+        }
+      } catch (JsonProcessingException e) {
+        log.error("Unable to send auto-accepted annotation for updated specimen: {}",
+            updatedDigitalSpecimenRecord.digitalSpecimenRecord().id(), e);
+      }
+    }
+  }
+
+
+  private List<AnnotationProcessingRequest> convertJsonPatchToAnnotations(
+      DigitalSpecimenRecord digitalSpecimenRecord, JsonNode jsonNode)
+      throws JsonProcessingException {
+    var annotations = new ArrayList<AnnotationProcessingRequest>();
+    var sourceSystemID = digitalSpecimenRecord.digitalSpecimenWrapper().attributes()
+        .getOdsSourceSystemID();
+    var sourceSystemName = digitalSpecimenRecord.digitalSpecimenWrapper().attributes()
+        .getOdsSourceSystemName();
+    for (JsonNode action : jsonNode) {
+      var annotationProcessingRequest = new AnnotationProcessingRequest()
+          .withOaHasTarget(buildTarget(digitalSpecimenRecord, buildSpecimenSelector(action)))
+          .withDctermsCreated(Date.from(Instant.now()))
+          .withDctermsCreator(
+              new Agent().withType(Type.AS_APPLICATION).withId(sourceSystemID)
+                  .withSchemaName(sourceSystemName));
+      if (action.get("op").asText().equals("replace")) {
+        annotationProcessingRequest.setOaMotivation(OaMotivation.OA_EDITING);
+        annotationProcessingRequest.setOaMotivatedBy(
+            "Received update information from Source System with id: " + sourceSystemID);
+        annotationProcessingRequest.setOaHasBody(
+            buildBody(extractValueString(action), sourceSystemID));
+      } else if (action.get("op").asText().equals("add")) {
+        annotationProcessingRequest.setOaMotivation(OaMotivation.ODS_ADDING);
+        annotationProcessingRequest.setOaMotivatedBy(
+            "Received new information from Source System with id: " + sourceSystemID);
+        annotationProcessingRequest.setOaHasBody(
+            buildBody(extractValueString(action), sourceSystemID));
+      } else if (action.get("op").asText().equals("remove")) {
+        annotationProcessingRequest.setOaMotivation(OaMotivation.ODS_DELETING);
+        annotationProcessingRequest.setOaMotivatedBy(
+            "Received delete information from Source System with id: " + sourceSystemID);
+      }
+      annotations.add(annotationProcessingRequest);
+    }
+    return annotations;
+  }
+
+  private String extractValueString(JsonNode action) throws JsonProcessingException {
+    if (action.get(VALUE).isTextual()) {
+      return action.get(VALUE).textValue();
+    } else {
+      return mapper.writeValueAsString(action.get(VALUE));
+    }
+  }
+
+  private OaHasSelector buildSpecimenSelector(JsonNode action) {
+    var pointer = action.get("path").asText();
+    var path = convertJsonPointToJsonPath(pointer);
+    if (action.get("path").asText().endsWith("/-")) {
```

Comment:
I couldn't find anything on a jsonPointer ending with "/-", what does this imply?

Reply:
"If the "-" character is used to index the end of the array (see [[RFC6901](https://datatracker.ietf.org/doc/html/rfc6901)]), this has the effect of appending the value to the array." So basically it indicates that it wasn't to add an object to an array. In the test I used the example of an EntityRelationship.